Converted TextMap propagator getter to a class and added keys method

nprajilesh opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #1084

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit Tests

# Checklist:

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

lzchen commented on opentelemetry-sdk/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 - Allow for Custom Trace and Span IDs Generation - `IdsGenerator` for TracerProvider
   ([#1153](https://github.com/open-telemetry/opentelemetry-python/pull/1153))
+- Add keys method to TextMap propagator Getter
+  ([#1084](https://github.com/open-telemetry/opentelemetry-python/issues/1084))
```
```suggestion
  ([#1196](https://github.com/open-telemetry/opentelemetry-python/issues/1196))
```

lzchen commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -32,18 +32,32 @@
 
-def get_header_from_scope(scope: dict, header_name: str) -> typing.List[str]:
-    """Retrieve a HTTP header value from the ASGI scope.
+class Getter:
```
Shouldn't this extend from `Getter` in the api package?

lzchen commented on opentelemetry-api/src/opentelemetry/baggage/propagation/__init__.py:
```diff
@@ -29,7 +29,7 @@ class BaggagePropagator(textmap.TextMapPropagator):
     def extract(
         self,
-        get_from_carrier: textmap.Getter[textmap.TextMapPropagatorT],
+        get_from_carrier: textmap.Getter,
```
It doesn't seem like `Getter` has an implementation?

lzchen commented on exporter/opentelemetry-exporter-datadog/tests/test_datadog_format.py:
```diff
@@ -22,9 +22,18 @@
-def get_as_list(dict_object, key):
-    value = dict_object.get(key)
-    return [value] if value is not None else []
+class Getter:
```
Shouldn't we be testing the actual implementation of the `Getter` used instead of a differently defined one in the test?

owais commented on instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:
```diff
@@ -248,8 +249,14 @@ def _trace_retry(*args, **kwargs):
+
+    @staticmethod
+    def keys(carrier):
+        return carrier.keys()
```
This feels a bit too verbose just to satisfy an interface. Can we have a convenience function that creates one for us? Something like:

```python
class Getter:
    def __init__(self, get=default_get, keys=default_keys):
        self._get = get
        self._keys = keys

    def get(self, carrier, key):
         return self._get(carrier, key)

    def keys(self, carrier):
         return self._keys(carrier)
```

Then auto and manual instrumented code can use it like:

```
getter = Getter(get=carrier_extractor)
```

This should reduce some boilerplate and allow users to just pass an inline lambda function if that'll get the job done instead of writing the whole class with both methods.

nprajilesh replied:
@owais Implemented this and added comments below

nprajilesh commented:
@lzchen  @owais I made code changes to implement it in the above-suggested way but I'm facing some typing errors. 


Currently, we have carrier defined as of type [TextMapPropagatorT](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py#L20) with no type constraints. Running a get() or keys() on this type throws typing errors because those 2 methods are defined for type Dict.

details of build failure - https://github.com/open-telemetry/opentelemetry-python/pull/1196/checks?check_run_id=1231130288

should we constrain the type of TextMapPropagator in the definition? I tried doing this but its throwing errors across multiple files that use the propagator.

lzchen commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,64 @@
+    return list(carrier.keys())
+
+
+GetterGetFunction = typing.Callable[
```
Nit: List the types at the top with `Setter`.

lzchen commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,64 @@
+GetterGetFunction = typing.Callable[
+    [TextMapPropagatorT, str], typing.List[str]
+]
+GetterKeysFunction = typing.Callable[[TextMapPropagatorT], typing.List[str]]
```
Having `Function` as a suffix is misleading as these are just types. Maybe just `GetterKeys` and `GetterGet`.

lzchen commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,64 @@
+        Returns:
+            first value of the propagation key or an empty list if the key doesn't exist.
+        """
+        return self._get(carrier, key)
```
I believe `get` is supposed to return `None` if not found. https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#get

nprajilesh replied:
@lzchen the existing get_from_carrier method is returning an empty list and most of the implementations of get_from_carrier are using an empty list.
https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py#L51 

Should I still make this change?

owais replied:
Yes, I think most instrumentations rely on it returning an empty list but we should change it if the spec says it should return None. I suggest to do that in a separate PR if you want to take it up @nprajilesh

nprajilesh replied:
@owais Sure, I can pick it up. I would prefer closing this before making that change.

I just have that one typing error remaining in this, If you could help me with that I will be done with this.

lzchen replied:
If you can make an issue for that, that'd be great.

lzchen replied:
https://github.com/open-telemetry/opentelemetry-python/issues/1274

lzchen replied:
@nprajilesh 
Would you like to take this one on?

lzchen commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,64 @@
-Getter = typing.Callable[[TextMapPropagatorT, str], typing.List[str]]
+
+
+def default_get(carrier: TextMapPropagatorT, key: str) -> typing.List[str]:
```
Why do we have default implementations for `get` and `keys`? Where is this implementation from?

nprajilesh replied:
this is implemented as @owais suggested, https://github.com/open-telemetry/opentelemetry-python/pull/1196#discussion_r501678110


and implementation of default function is taken as this was the implementation in most of the files.


I'm facing a typing issue with TextMapPropagatorT type, Could you please help with this, details can be found in this comment 
https://github.com/open-telemetry/opentelemetry-python/pull/1196#issuecomment-706678858

owais replied:
Took a quick look. Your implementation is accepting a TextMapPropagatorT type but then using it like a dictionary resulting in the error. Do we need a default implementation here? Suggest removing it and keeping only interfaces here. The SDK package may contain a carrier specific implementation (one for dictionaries for example) if a common one is used across the project.

nprajilesh replied:
@owais If we remove the default implementation, all the files using getter will have to implement keys even if they don't use it.

owais replied:
To fix your typing errors, they should be implementation specific. AFAIU, what we need is a dict implementation of getter as that is used almost everywhere. Such an implementation should be specific to dictionaries, i.e, accept dictionaries as the carrier argument, not the generic var.

lzchen replied:
@owais 
Should we keep the default implementations then? (common across many instrumentations)

owais replied:
Yes, we can keep it. My comment was more about keeping it in API vs moving to SDK. I'm fine with either.

nprajilesh replied:
> Took a quick look. Your implementation is accepting a TextMapPropagatorT type but then using it like a dictionary resulting in the error. Do we need a default implementation here? Suggest removing it and keeping only interfaces here. The SDK package may contain a carrier specific implementation (one for dictionaries for example) if a common one is used across the project.

@owais I'm a bit confused with this one, here is my understanding

- If we are going to remove the default implementation we have to implement get() and keys() everywhere. 
If we go with keeping the default implementation in textmap.py we will have a typing error because the default implementation uses a dict type and carrier expects a type of generic var(TextMapPropagatorT). If we have to change the type of the get method of getter to  `typing.Callable[[typing.Dict[str,str], str], typing.List[str]]` we have to change the carrier type to Dict as well from generic var in all the implementations. 

what do you suggest we do?

owais replied:
May be it is stemming from my little understanding of mypy but I think there are two issues overlapping here. 

1. Whether and where to keep a commonly used implementation for dicts as dicts seem to be most common carriers. Perhaps we shouldn't call it default implementation, just a dicts one. 
2. My original comment was about reducing boilerplate code in instrumentations and user code. The suggestion was to have a function or utility class that could turn simple lambdas or functions into fully compatible Getter implementations so that each instrumentation did not have to define a class and all methods to implement a getter. If  dict is being used in 99% of the cases and we just need to add custom getter implementations for in just one or two places then feel free to not add this.

I have a working example below. It defines the Getter interface and then implements two Getter implementations for List and Dict types. It also implements a "Helper" implementation (which I had originally requested) that can be used to turn any set of get and keys functions into a compatible getter. 

```python
from typing import TypeVar, Generic, List, Dict, Callable


CarrierT = TypeVar('CarrierT')


class Getter(Generic[CarrierT]):
    def get(self, carrier: CarrierT, key: str) -> List[str]:
        raise NotImplementedError()
    
    def keys(self, carrier: CarrierT) -> List[str]:
        raise NotImplementedError()
    

def print_carrier_contents(getter: Getter, carrier: CarrierT):
    for key in getter.keys(carrier):
        print(key, ' = ', getter.get(carrier, key))
    print()


try:
   print_carrier_contents(Getter(), {})
except NotImplementedError:
   pass

class ListGetter(Getter):
    def get(self, carrier: List[str], key: str) -> List[str]:
        for pair in carrier:
            if pair[0] == key:
                return [pair[1]]
        return []
    
    def keys(self, carrier: List[str]) -> List[str]:
        return [i[0] for i in carrier]
    
print_carrier_contents(ListGetter(), [("k1", "v1"), ("k2", "v2")])


class DictGetter(Getter):
    def get(self, carrier: Dict[str, str], key: str) -> List[str]:
        val = carrier.get(key, None)
        if val:
          return [val]
        return []
    
    def keys(self, carrier: Dict[str, str]) -> List[str]:
        return list(carrier.keys())

print_carrier_contents(DictGetter(), {"k1": "v1", "k2": "v2"})



class HelperGetter(Getter):

    def __init__(self, get: Callable[[CarrierT, str], List[str]], keys: Callable[[CarrierT], List[str]]):
      self._get = get
      self._keys = keys

    def get(self, carrier: CarrierT, key: str) -> List[str]:
      return self._get(carrier, key)

    def keys(self, carrier: CarrierT) -> List[str]:
      return self._keys(carrier)


class CustomCarrier:
    
    def get_value(self, key):
        if key == "k1":
            return ["v1"]
        elif key == "k2":
            return ["v2"]
        return []
        
    def list_keys(self):
        return ["k1", "k2"]
        
        
getter = HelperGetter(get=CustomCarrier.get_value, keys=CustomCarrier.list_keys)

print_carrier_contents(getter, CustomCarrier())
```

Hope this helps.

NathanielRN commented:
Hello! I have a PR to move some files you have in this PR to the Contrib repo, please let me know if this gets merged before the PR in the Contrib repo. Please see https://github.com/open-telemetry/opentelemetry-python-contrib/pulls/75

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,68 @@
+    def get(
+        self, carrier: typing.Dict[str, str], key: str
+    ) -> typing.List[str]:
+        val = carrier.get(key, None)
```
This should check if the value is already a list or collection and return it as is but if the value is a single item, then return it as `[val]`

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -20,7 +20,68 @@
+        return list(carrier.keys())
+
+
+class HelperGetter(Getter[TextMapPropagatorT]):
```
This is a very bad name. I didn't mean to suggest it. Was trying to conjure up a working example real quick so didn't think of a better name. I still don't have a better suggestion but it for sure shouldn't be "HelperGetter" IMO :) 

How about `CustomGetter` since it takes custom implementations for get and keys methods?

nprajilesh replied:
sure, this make sense

owais commented on opentelemetry-api/tests/baggage/test_baggage_propagation.py:
```diff
@@ -26,14 +27,17 @@ def get_as_list(
     return dict_object.get(key, [])
 
 
+getter = HelperGetter(get_as_list, DictGetter.keys)
```
I think we should use DictGetter directly for such cases. We just need to modify DictGetter to work with both of the following types of dictionaries:

1. `Dict[str, str]`
2. `Dict[str, Collection[str]]`

DictGetter should work with both of these cases. Once we support that, we can use dict getter directly in places like these.

nprajilesh replied:
I've added this removed duplicate get implementations in other files

owais commented on exporter/opentelemetry-exporter-datadog/src/opentelemetry/exporter/datadog/propagator.py:
```diff
@@ -44,19 +44,19 @@ def extract(
     ) -> Context:
         trace_id = extract_first_element(
-            get_from_carrier(carrier, self.TRACE_ID_KEY)
+            get_from_carrier.get(carrier, self.TRACE_ID_KEY)
```
`get_from_carrier` should be renamed to `carrier_getter` or `getter`.

owais commented on exporter/opentelemetry-exporter-datadog/tests/test_datadog_format.py:
```diff
@@ -18,13 +18,11 @@
-def get_as_list(dict_object, key):
-    value = dict_object.get(key)
-    return [value] if value is not None else []
+getter = DictGetter()
```
perhaps `carrier_getter` would be a better name for such a global var so it's more obvious when reading code way below this line.

owais commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -46,6 +47,9 @@ def get_header_from_scope(scope: dict, header_name: str) -> typing.List[str]:
     ]
 
 
+getter = HelperGetter(get_header_from_scope, DictGetter.keys)
```
We probably never call `getter.keys()` on this instance but would it work if something did call it? I guess it would either complain that it recieved one less argument (self) or it would pass instance of "HelperGetter" as the value for "self" which may cause some very hard to debug bugs down the line. 

Also, since the actual getter function being used here is operating on dict, any reason to not just sub-class DictGetter and override the `get()` method?

owais commented on instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:
```diff
@@ -118,7 +119,8 @@ def _trace_prerun(self, *args, **kwargs):
 
         request = task.request
-        tracectx = propagators.extract(carrier_extractor, request) or None
+        getter = HelperGetter(carrier_extractor, DictGetter.keys)
```
1. getter should be initialized once per module as they are stateless and don't need to be re-created for every span.
2. if something ever called `getter.keys()` on this instance, it would panic because `task.request` is most-likely not a dictionary according to the implementation of `carrier_extractor`. It looks like it is an object and carrier_extractor is using getattr to extract carrier values from object attributes. I'm not sure what an ideal implementation of keys would be here. It may be to have it as noop so it always returns an empty list or it may be to return every attribute the carrier has. A third option might be to return well-known carrier keys when present on the object as attributes.

nprajilesh replied:
@owais I've updated the getter as subclass of DictGetter and overridden the keys method to  return an empty list

owais commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:
```diff
@@ -126,8 +127,9 @@ def _set_remote_context(self, servicer_context):
             def get_from_grpc_metadata(metadata, key) -> List[str]:
                 return [md_dict[key]] if key in md_dict else []
 
+            getter = HelperGetter(get_from_grpc_metadata, DictGetter.keys)
```
should be initialized once either in init or globally for the module.

owais reviewed (commented):
Thanks for updating this @nprajilesh. Sorry about the initial confusion. I think we are very close to merging this. Left a few small comments on usage.

nprajilesh commented:
@owais Thank you for all the suggestions and review comments!  I've made the suggested changes, 

- updated DictGetter to support str, List[str]
- Modified getter to be a subclass wherever necessary
- removed duplicate implementations of get
- updated the variable/class names as suggested

owais commented on instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py:
```diff
@@ -61,24 +61,31 @@ def hello():
-    environ: dict, header_name: str
-) -> typing.List[str]:
-    """Retrieve a HTTP header value from the PEP3333-conforming WSGI environ.
+# pylint:disable=arguments-differ
```
Is this because we renamed carrier to environ and key to header_name in the subclass? If so, why not use carrier and key in the first place? It's anyway safer in case something calls the getter with named arguments.

nprajilesh replied:
the initial get function which was used in WSGI had these argument names, I consider changing it to carrier and key but this made more contextual sense.

owais replied:
I think we should stick to argument names specified by the getter interface. I feel using different names isn't adding enough value to justify not satisfying the interface 100%. We can explain in doc string that carrier is WSGI environ and key is header name.

owais replied:
@nprajilesh how do you think we should handle this?

nprajilesh replied:
@owais yah, I think you are right. I've updated the argument names and added docstring for clarity.

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -18,9 +18,71 @@
+        or more values from the carrier. In the case that
+        the value does not exist, returns an empty list.
+
+        Args: carrier: and object which contains values that are used to
```
should this be on the next line?

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -18,9 +18,71 @@
+        Args: carrier: and object which contains values that are used to
+        construct a Context. This object must be paired with an appropriate
+        getter which understands how to extract a value from it.
+        key: key of a field in carrier. Returns: first value of the
```
new line before return values

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -18,9 +18,71 @@
+        val = carrier.get(key, None)
+        if not val:
+            return []
+        return val if isinstance(val, typing.List) else [val]
```
would it  be better to have this:

```python
    if not isinstance(val, str) and isinstance(value, Iterable):
        return list(val)
    return [val]
```

This way tuples or any other collection values would work as well.

owais commented on opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py:
```diff
@@ -18,9 +18,71 @@
+        return list(carrier.keys())
+
+
+class CustomGetter(Getter[TextMapPropagatorT]):
```
Looks like this isn't used anywhere except one test case after all. Let's remove it if it turns out that we don't need it.

owais reviewed (commented):
Looks great! Left some final comments.

owais commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -29,21 +29,22 @@
-    """Retrieve a HTTP header value from the ASGI scope.
+# pylint:disable=arguments-differ
+class CarrierGetter(DictGetter):
+    def get(self, scope: dict, header_name: str) -> typing.List[str]:
```
nit: should use carrier and key argument names and if you want, use doc string to explain that they are supposed to be asgi scope and header name.

owais replied:
@nprajilesh ^

owais reviewed (commented):
Looks good. Just need two remaining comments to be resolved.

lzchen commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -29,21 +29,31 @@
 
-def get_header_from_scope(scope: dict, header_name: str) -> typing.List[str]:
-    """Retrieve a HTTP header value from the ASGI scope.
+class CarrierGetter(DictGetter):
```
Shouldn't these be extending from `Getter` instead of `DictGetter`?

nprajilesh replied:
@lzchen Extending DictGetter because of the following reasons:

- carrier is of type dict
- there is no keys implementation required here so we can use the default keys implementation in DictGetter
- if we have to extend the Getter Implementation we have to override the keys() method as well

owais replied:
Yea, this getter is more of a specialized implementation of a dict getter so LGTM.

nprajilesh commented:
@codeboten I have resolved the conflicts.

codeboten reviewed (approved):
This is a big improvement in the code, thanks for implementing it!
